sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventHubConsumer.cs

```diff
@@ -130,11 +161,111 @@ public override void UpdateRetryPolicy(EventHubRetryPolicy newRetryPolicy)
+
+            var stopWatch = Stopwatch.StartNew();
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        EventHubsEventSource.Log.EventReceiveStart(EventHubName, ConsumerGroup, PartitionId);
+
+                        link = await ReceiveLink.GetOrCreateAsync(UseMinimum(ConnectionScope.SessionTimeout, _tryTimeout)).ConfigureAwait(false);
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        var messagesReceived = await Task.Factory.FromAsync
+                        (
+                            (callback, state) => link.BeginReceiveMessages(maximumMessageCount, waitTime, callback, state),
+                            (asyncResult) => link.EndReceiveMessages(asyncResult, out amqpMessages),
+                            TaskCreationOptions.RunContinuationsAsynchronously
+                        ).ConfigureAwait(false);
+
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        // If event messages were received, then package them for consumption and
+                        // return them.
+
+                        if ((messagesReceived) && (amqpMessages != null))
+                        {
+                            receivedEvents ??= new List<EventData>();
+
+                            foreach (AmqpMessage message in amqpMessages)
+                            {
+                                link.DisposeDelivery(message, true, AmqpConstants.AcceptedOutcome);
+                                receivedEvents.Add(MessageConverter.CreateEventFromMessage(message));
+                                message.Dispose();
+                            }
+
+                            receivedEventCount = receivedEvents.Count;
+
+                            if (LastEnqueuedEventInformation != null)
+                            {
+                                EventData lastEvent = receivedEvents[receivedEventCount - 1];
+                                LastEnqueuedEventInformation.UpdateMetrics(lastEvent.LastPartitionSequenceNumber, lastEvent.LastPartitionOffset, lastEvent.LastPartitionEnqueuedTime, DateTimeOffset.UtcNow);
+                            }
+
+                            return receivedEvents;
+                        }
+
+                        // No events were available.
+
+                        return Enumerable.Empty<EventData>();
+                    }
+                    catch (EventHubsTimeoutException)
+                    {
+                        // Because the timeout specified with the request is intended to be the maximum
+                        // amount of time to wait for events, a timeout isn't considered an error condition,
+                        // rather a sign that no events were available in the requested period.
```

Comment:
What if we get a timeout when creating the link?

Reply:
That doesn't bubble up as an Event Hubs timeout; that comes as a `System.Timeout` exception, which would bypass the `catch` on 230 and enter standard exception handling/retry logic on 238.

_(see: [TimeoutAsyncResult](https://github.com/Azure/azure-amqp/blob/4dcaf85d6073902c62b6866112447020321095da/src/TimeoutAsyncResult.cs#L37), line 37)_